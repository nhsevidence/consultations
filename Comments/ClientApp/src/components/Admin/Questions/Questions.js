import React, {Component, Fragment} from "react";

export class Questions extends Component {
	render(){
		return  (
			<Fragment>
				<h2>Questions</h2>
				<div className="grid">
					<div data-g="12 md:3">
						left
					</div>
					<div data-g="12 md:9">
						right
					</div>
				</div>
			</Fragment>
		);
	}
}
