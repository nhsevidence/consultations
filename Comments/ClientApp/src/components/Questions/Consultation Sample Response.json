{
  "reference": "GID-NG10109",
  "title": "Consultation Comments Test project (Numbered Paragraphs )",
  "consultationName": "Draft guidance consultation",
  "startDate": "2018-03-01T00:00:00",
  "endDate": "2056-12-24T17:00:00",
  "consultationType": "Draft guidance consultation",
  "resourceTitleId": null,
  "projectType": "NG",
  "productTypeName": "NICE guideline",
  "developedAs": null,
  "relevantTo": null,
  "consultationId": 22,
  "process": "NG",
  "hasDocumentsWhichAllowConsultationComments": false,
  "hasDocumentsWhichAllowConsultationQuestions": true,
  "supportsQuestions": true,
  "supportsComments": true,
  "partiallyUpdatedProjectReference": null,
  "origProjectReference": null,
  "user": {
    "isAuthorised": true,
    "displayName": "Warren Keith",
    "userId": "0b09270a-7a30-4cdb-86e6-7d19d0cea1ba",
    "organisationName": null
  },
  "consultationState": {
    "startDate": "2018-03-01T00:00:00",
    "endDate": "2056-12-24T17:00:00",
    "hasQuestions": true,
    "hasUserSuppliedAnswers": true,
    "hasUserSuppliedComments": true,
    "userHasSubmitted": false,
    "consultationSupportsQuestions": false,
    "consultationSupportsComments": false,
    "documentIdsWhichSupportQuestions": [
      1,
      2,
      3,
      4
    ],
    "documentIdsWhichSupportComments": [
      1,
      2,
      3,
      4
    ],
    "consultationIsOpen": true,
    "consultationHasNotStartedYet": false,
    "consultationHasEnded": false,
    "supportsSubmission": true,
    "supportsDownload": true,
    "shouldShowDrawer": true,
    "shouldShowCommentsTab": true,
    "shouldShowQuestionsTab": true
  },
  "breadcrumbs": [
    {
      "label": "Home",
      "url": "/",
      "localRoute": false
    },
    {
      "label": "Consultation Comments Test project (Numbered Paragraphs )",
      "url": "/guidance/indevelopment/gid-ng10109/consultation/html-content",
      "localRoute": false
    }
  ],
  "filters": null
}
