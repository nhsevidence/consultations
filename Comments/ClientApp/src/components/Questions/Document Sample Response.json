[
  {
    "consultationId": 22,
    "documentId": 1,
    "convertedDocument": true,
    "supportsComments": true,
    "supportsQuestions": true,
    "title": "Unstable angina and NSTEMI",
    "chapters": [
      {
        "slug": "introduction",
        "title": "Introduction"
      },
      {
        "slug": "patient-centred-care",
        "title": "Patient-centred care"
      },
      {
        "slug": "key-priorities-for-implementation",
        "title": "Key priorities for implementation"
      },
      {
        "slug": "guidance",
        "title": "1 Guidance"
      },
      {
        "slug": "notes-on-the-scope-of-the-guidance",
        "title": "2 Notes on the scope of the guidance"
      },
      {
        "slug": "implementation",
        "title": "3 Implementation"
      },
      {
        "slug": "research-recommendations",
        "title": "4 Research recommendations"
      },
      {
        "slug": "other-versions-of-this-guideline",
        "title": "5 Other versions of this guideline"
      },
      {
        "slug": "related-nice-guidance",
        "title": "6 Related NICE guidance"
      },
      {
        "slug": "updating-the-guideline",
        "title": "7 Updating the guideline"
      },
      {
        "slug": "appendix-a-the-guideline-development-group-and-nice-project-team",
        "title": "Appendix A: The Guideline Development Group and NICE project team"
      },
      {
        "slug": "appendix-b-the-guideline-review-panel",
        "title": "Appendix B: The Guideline Review Panel"
      },
      {
        "slug": "appendix-c-the-algorithm",
        "title": "Appendix C: The algorithm"
      },
      {
        "slug": "changes-after-publication",
        "title": "Changes after publication"
      },
      {
        "slug": "about-this-guideline",
        "title": "About this guideline"
      }
    ],
    "href": "/guidance/GID-NG10109/documents/129",
    "sourceURI": "consultations://./consultation/22/document/1"
  },
  {
    "consultationId": 22,
    "documentId": 2,
    "convertedDocument": true,
    "supportsComments": true,
    "supportsQuestions": true,
    "title": "Glaucoma: diagnosis and management – Full Guideline",
    "chapters": [
      {
        "slug": "glaucoma-diagnosis-and-management-full-guideline",
        "title": "Glaucoma: diagnosis and management – Full Guideline"
      },
      {
        "slug": "acknowledgements",
        "title": "Acknowledgements"
      },
      {
        "slug": "guideline-summary",
        "title": "1 Guideline summary"
      },
      {
        "slug": "introduction",
        "title": "2 Introduction"
      },
      {
        "slug": "development-of-the-guideline",
        "title": "3 Development of the guideline"
      },
      {
        "slug": "methods",
        "title": "4 Methods"
      },
      {
        "slug": "prognostic-risk-tools",
        "title": "5 Prognostic risk tools"
      },
      {
        "slug": "tests-used-in-case-finding-diagnosis-and-reassessment",
        "title": "6 Tests used in case finding, diagnosis and reassessment"
      },
      {
        "slug": "reassessment-intervals",
        "title": "7 Reassessment intervals"
      },
      {
        "slug": "overview-of-treatment",
        "title": "8 Overview of treatment"
      },
      {
        "slug": "treatment-of-ocular-hypertension-suspected-chronic-open-angle-glaucoma-and-confirmed-chronic",
        "title": "9 Treatment of ocular hypertension, suspected chronic open-angle glaucoma and confirmed chronic open-angle glaucoma"
      },
      {
        "slug": "complementary-and-alternative-interventions",
        "title": "10  Complementary and alternative interventions"
      },
      {
        "slug": "organisation-of-care",
        "title": "11 Organisation of care"
      },
      {
        "slug": "provision-of-information-for-patients",
        "title": "12 Provision of information for patients"
      },
      {
        "slug": "reference-list",
        "title": "13 Reference list"
      },
      {
        "slug": "acronyms-and-abbreviations",
        "title": "14 Acronyms and abbreviations"
      },
      {
        "slug": "glossary",
        "title": "15 Glossary"
      }
    ],
    "href": "/guidance/GID-NG10109/documents/129-2",
    "sourceURI": "consultations://./consultation/22/document/2"
  },
  {
    "consultationId": 22,
    "documentId": 0,
    "convertedDocument": false,
    "supportsComments": false,
    "supportsQuestions": false,
    "title": "Comments form",
    "chapters": null,
    "href": "/guidance/GID-NG10109/documents/comments-form",
    "sourceURI": null
  },
  {
    "consultationId": 22,
    "documentId": 3,
    "convertedDocument": true,
    "supportsComments": true,
    "supportsQuestions": true,
    "title": "Emergency and acute medical care in over 16s: service delivery and organisation",
    "chapters": [
      {
        "slug": "overview",
        "title": "Overview"
      },
      {
        "slug": "context",
        "title": "Context"
      },
      {
        "slug": "recommendations",
        "title": "Recommendations"
      },
      {
        "slug": "putting-this-guideline-into-practice",
        "title": "Putting this guideline into practice"
      },
      {
        "slug": "recommendations-for-research-3",
        "title": "Recommendations for research"
      }
    ],
    "href": "/guidance/GID-NG10109/documents/129-3",
    "sourceURI": "consultations://./consultation/22/document/3"
  },
  {
    "consultationId": 22,
    "documentId": 0,
    "convertedDocument": false,
    "supportsComments": false,
    "supportsQuestions": false,
    "title": "Addendum appendix",
    "chapters": null,
    "href": "/guidance/GID-NG10109/documents/addendum-appendix",
    "sourceURI": null
  },
  {
    "consultationId": 22,
    "documentId": 4,
    "convertedDocument": true,
    "supportsComments": true,
    "supportsQuestions": true,
    "title": "Nivolumab for adjuvant treatment of completely resected melanoma with lymph node involvement or metastatic disease",
    "chapters": [
      {
        "slug": "recommendations",
        "title": "1 Recommendations"
      },
      {
        "slug": "information-about-nivolumab",
        "title": "2 Information about nivolumab"
      },
      {
        "slug": "committee-discussion",
        "title": "3 Committee discussion"
      },
      {
        "slug": "proposed-date-for-review-of-guidance",
        "title": "4 Proposed date for review of guidance"
      },
      {
        "slug": "appraisal-committee-members-and-nice-project-team",
        "title": "5 Appraisal committee members and NICE project team"
      }
    ],
    "href": "/guidance/GID-NG10109/documents/129-4",
    "sourceURI": "consultations://./consultation/22/document/4"
  }
]
