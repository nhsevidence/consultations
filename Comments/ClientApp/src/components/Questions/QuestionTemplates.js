import React, { Component, Fragment } from "react";

// HSCTeam
// CfGTeam
// CHTETeam

export class QuestionTemplates extends Component {
	state = {
		filter: {
			string: "",
			directorate: true,
		},
	};

	handleFilter = (event, filterByDirectorate) => {
		if (filterByDirectorate === null) {
			this.setState({
				filter: {
					string: event.target.value,
					directorate: this.state.filter.directorate,
				},
			});
		} else {
			this.setState({
				filter: {
					string: this.state.filter.string,
					directorate: filterByDirectorate,
				},
			});
		}
	};

	filteredQuestions = () => {
		return this.props.questions.filter(question =>
			question.questionText.toLowerCase().indexOf(this.state.filter.string.toLowerCase()) !== -1
			&&
			question.allRoles // todo: carry on here!
		);
	};

	render() {
		const {currentConsultationId, currentDocumentId, newQuestion} = this.props;
		const filteredQuestions = this.filteredQuestions();
		return (
			<div className="card">
				<h3>Previously set questions</h3>
				<QuestionsFilter
					handleFilter={this.handleFilter}
					filter={this.state.filter}/>
				{filteredQuestions.length === 0 ? <p><i>No matches</i></p> :
					<ul className="list--unstyled">
						{filteredQuestions.map(item =>
							(<TemplateItem
								{...item}
								key={item.questionId}
								currentConsultationId={currentConsultationId}
								currentDocumentId={currentDocumentId}
								newQuestion={newQuestion}/>))}
					</ul>
				}
			</div>
		);
	}
}

const TemplateItem = (props) => {
	const {
		questionText,
		newQuestion,
		currentDocumentId,
		currentConsultationId,
		questionTypeId,
		questionType,
		allRoles,
	} = props;
	const getQuestionType = (type) => {
		return type === "Text" ? "Open Question" : "Yes/No Question";
	};
	return (
		<li>
			<div className="card">
				<div className="grid">
					<div data-g="9">
						<div className="card__header">
							<p className="card__heading">
								{questionText}
							</p>
						</div>
						<dl className="card__metadata">
							<div className="card__metadatum">
								<dt>
									<span className="card__tag tag tag--flush">
										{getQuestionType(questionType.type)}
									</span>
								</dt>
								<dd className="visually-hidden">{questionType.type}</dd>
							</div>
							{allRoles.map(role =>
								<Fragment key={role}>
									<div className="card__metadatum">
										<dt>
											<span className="card__tag tag tag--flush">
												{role}
											</span>
										</dt>
										<dd className="visually-hidden">{role}</dd>
									</div>
								</Fragment>,
							)}
						</dl>
					</div>
					<div data-g="3">
						<button
							onClick={e => {
								if (currentDocumentId === "consultation") {
									newQuestion(e, currentConsultationId, null, questionTypeId, questionText);
								} else {
									newQuestion(e, currentConsultationId, parseInt(currentDocumentId, 10), questionTypeId, questionText);
								}
							}}
							className="btn btn-small right mr--0">Insert
						</button>
					</div>
				</div>
			</div>
		</li>
	);
};

const QuestionsFilter = props => {
	const {handleFilter, filter} = props;
	return (
		<div className="panel mb--d">
			<div className="form__group form__group--text">
				<label htmlFor="textFilter" className="form__label"><b>Filter by question text</b></label>
				<input
					value={filter.string}
					className="form__input"
					onChange={e => handleFilter(e, null)}
					id="textFilter"
					tabIndex={0}/>
			</div>
			<div role="radiogroup">
				<label htmlFor="filterByRole" className="form__label"><b>Filter by directorate</b></label>
				<div className="form__group form__group--radio form__group--inline">
					<input
						className="form__radio"
						id="filterByRole--mine"
						type="radio"
						name="filterByRole"
						checked={filter.directorate === true}
						onChange={e => handleFilter(e, true)}
						value={true}
					/>
					<label
						className="form__label form__label--radio"
						htmlFor="filterByRole--mine">
						My Directorate
					</label>
				</div>
				<div className="form__group form__group--radio form__group--inline">
					<input
						className="form__radio"
						id="filterByRole--all"
						type="radio"
						name="filterByRole"
						checked={filter.directorate === false}
						onChange={e => handleFilter(e, false)}
						value={false}
					/>
					<label
						className="form__label form__label--radio"
						htmlFor="filterByRole--all">
						All Directorates
					</label>
				</div>
			</div>
		</div>
	);
};
