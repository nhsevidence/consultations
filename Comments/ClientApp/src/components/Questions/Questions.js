import React, { Component, Fragment } from "react";
import { withRouter } from "react-router-dom";
import Helmet from "react-helmet";

import { LoginBanner } from "../LoginBanner/LoginBanner";
import { Breadcrumbs } from "../Breadcrumbs/Breadcrumbs";
import { NestedStackedNav } from "../NestedStackedNav/NestedStackedNav";
import { UserContext } from "../../context/UserContext";
import preload from "../../data/pre-loader";
import { load } from "../../data/loader";

type PropsType = any; // todo

type StateType = any; // todo

export class Questions extends Component<PropsType, StateType> {
	constructor(props: PropsType) {
		super(props);

		this.state = {
			consultationData: null,
			loading: true,
			hasInitialData: false,
			error: {
				hasError: false,
				message: null,
			},
		};

		let preloadedConsultation, preloadedDocuments;

		let preloadedData = {};
		if (this.props.staticContext && this.props.staticContext.preload) {
			preloadedData = this.props.staticContext.preload.data; //this is data from Configure => SupplyData in Startup.cs. the main thing it contains for this call is the cookie for the current user.
		}

		preloadedConsultation = preload(
			this.props.staticContext,
			"consultation",
			[],
			{consultationId: this.props.match.params.consultationId, isReview: false},
			preloadedData,
		);

		preloadedDocuments = preload(
			this.props.staticContext,
			"documents",
			[],
			{consultationId: this.props.match.params.consultationId},
			preloadedData,
		);

		if (preloadedConsultation && preloadedDocuments) {
			this.state = {
				consultationData: preloadedConsultation,
				documentsData: preloadedDocuments,
				loading: false,
				hasInitialData: true,
				error: {
					hasError: false,
					message: null,
				},
			};
		}

	}

	gatherData = async () => {
		const {consultationId} = this.props.match.params;

		const consultationData = load("consultation", undefined, [], {
			consultationId,
			isReview: false,
		})
			.then(response => response.data)
			.catch(err => {
				this.setState({
					error: {
						hasError: true,
						message: "consultationData " + err,
					},
				});
			});

		const documentsData = load("documents", undefined, [], {consultationId})
			.then(response => response.data)
			.catch(err => {
				this.setState({
					error: {
						hasError: true,
						message: "documentsData " + err,
					},
				});
			});

		return {
			consultationData: await consultationData,
			documentsData: await documentsData,
		};
	};

	componentDidMount() {
		if (!this.state.hasInitialData) {
			this.gatherData()
				.then(data => {
					this.setState({
						...data,
						loading: false,
						hasInitialData: true,
					});
				})
				.catch(err => {
					this.setState({
						error: {
							hasError: true,
							message: "gatherData in componentDidMount failed " + err,
						},
					});
				});
		}
	}

	createConsultationNavigation = (consultationData, documentsData, currentConsultationId, currentDocumentId) => {
		const supportsQuestions = document => document.supportsQuestions;

		const isCurrentRoute = (consultationId, documentId) => {
			return (consultationId === parseInt(currentConsultationId))
				&& (documentId === parseInt(currentDocumentId));
		};

		const documentsList = documentsData
			.filter(supportsQuestions)
			.map(consultationDocument => {
				return {
					title: consultationDocument.title,
					to: `/admin/questions/${consultationDocument.consultationId}/${consultationDocument.documentId}`,
					marker: 2,
					current: isCurrentRoute(
						consultationDocument.consultationId,
						consultationDocument.documentId),
				};
			}
			);

		return [
			{
				title: consultationData.title,
				to: `/admin/questions/${consultationData.consultationId}/0`,
				marker: 3,
				current: isCurrentRoute(consultationData.consultationId, 0),
				children: documentsList,
			},
		];
	};

	render() {
		if (!this.state.hasInitialData) return null;

		const fakeLinks = [
			{
				label: "Back to In Development",
				url: "/",
				localRoute: false,
			},
			{
				label: "Add questions",
				url: this.props.match.url,
				localRoute: true,
			},
		];

		const {consultationData, documentsData} = this.state;

		const currentDocumentId = this.props.match.params.documentId;
		const currentConsultationId =this.props.match.params.consultationId;

		return (
			<UserContext.Consumer>
				{(contextValue: any) => !contextValue.isAuthorised ?
					<LoginBanner
						signInButton={false}
						currentURL={this.props.match.url}
						signInURL={contextValue.signInURL}
						registerURL={contextValue.registerURL}
						signInText="to administer a consultation"
					/>
					:
					<Fragment>
						<Helmet>
							<title>Set Questions</title>
						</Helmet>
						<div className="container">
							<div className="grid">
								<div data-g="12">
									<Breadcrumbs links={fakeLinks}/>
									<h1 className="h3">{consultationData.title}</h1>
									<div className="grid">
										<div data-g="12 md:6">
											<NestedStackedNav
												navigationStructure={
													this.createConsultationNavigation(
														consultationData,
														documentsData,
														currentConsultationId,
														currentDocumentId,
													)
												}/>
										</div>
										<div data-g="12 md:6">
											{!currentDocumentId &&
												<p>Choose a consultation or document to add questions</p>
											}
										</div>
									</div>
								</div>
							</div>
						</div>
					</Fragment>
				}
			</UserContext.Consumer>
		);
	}
}

export default withRouter(Questions);
