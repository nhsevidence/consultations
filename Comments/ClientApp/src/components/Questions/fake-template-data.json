[
  {
    "_id": "5d1df8d7c9a51d6d8aacfac1",
    "question": "Do you agree with the proposal not to update the guideline?"
  },
  {
    "_id": "5d1df8d7ba80fa6498b5cc7c",
    "question": "Do you have any comments on areas excluded from the scope of the guideline?"
  },
  {
    "_id": "5d1df8d733eb5122a97d46ef",
    "question": "Do you have any comments on equalities issues?"
  },
  {
    "_id": "5d1df8d786b906d5e94788a6",
    "question": "Has all of the relevant evidence been taken into account?"
  },
  {
    "_id": "5d1df8d7e898e609e1b92834",
    "question": "Are the summaries of clinical and cost effectiveness reasonable interpretations of the evidence?"
  },
  {
    "_id": "5d1df8d7e898e609e1b928343",
    "question": "Are the recommendations sound and a suitable basis for guidance to the NHS?"
  },
  {
    "_id": "5d1df8d7e898e609e1b928344",
    "question": "Are there any aspects of the recommendations that need particular consideration to ensure we avoid unlawful discrimination against any group of people on the grounds of race, gender, disability, religion or belief, sexual orientation, age, gender reassignment, pregnancy and maternity?"
  }
]
