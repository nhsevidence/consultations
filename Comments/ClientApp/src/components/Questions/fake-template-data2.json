{
  "questions": [
    {
        "questionId": 4,
        "questionText": "Which areas will have the biggest impact on practice and be challenging to implement? Please say for whom and why.",
        "questionTypeId": 1,
        "questionOrder": 1,
        "lastModifiedByUserId": "00000000-0000-0000-0000-000000000000",
        "lastModifiedDate": "2018-08-03T22:44:58.45",
        "questionType": {
            "description": "A text question requiring a text answer.",
            "hasTextAnswer": true,
            "hasBooleanAnswer": false,
            "type":"Text"
        },
        "locationId": 7,
        "htmlElementID": null,
        "rangeStart": null,
        "rangeStartOffset": null,
        "rangeEnd": null,
        "rangeEndOffset": null,
        "quote": null,
        "commentOn": "Consultation",
        "documentId": null,
        "show": true
    },
    {
        "questionId": 5,
        "questionText": "Would implementation of any of the draft recommendations have significant cost implications?",
        "questionTypeId": 1,
        "questionOrder": 2,
        "lastModifiedByUserId": "00000000-0000-0000-0000-000000000000",
        "lastModifiedDate": "2018-08-03T22:44:58.45",
        "questionType": {
            "description": "A text question requiring a text answer.",
            "hasTextAnswer": true,
            "hasBooleanAnswer": false,
            "type":"Text"
        },
        "locationId": 8,
        "htmlElementID": null,
        "rangeStart": null,
        "rangeStartOffset": null,
        "rangeEnd": null,
        "rangeEndOffset": null,
        "quote": null,
        "commentOn": "Consultation",
        "documentId": 1
    },
    {
        "questionId": 6,
        "questionText": "Do you like rainbows?",
        "questionTypeId": 1,
        "questionOrder": 3,
        "lastModifiedByUserId": "00000000-0000-0000-0000-000000000000",
        "lastModifiedDate": "2018-08-03T22:44:58.45",
        "questionType": {
            "description": "A yes or no question.",
            "hasTextAnswer": true,
            "hasBooleanAnswer": true,
            "type":"YesNo"
        },
        "locationId": 9,
        "htmlElementID": null,
        "rangeStart": null,
        "rangeStartOffset": null,
        "rangeEnd": null,
        "rangeEndOffset": null,
        "quote": null,
        "commentOn": "Consultation",
        "documentId": 2
    }
  ]
}