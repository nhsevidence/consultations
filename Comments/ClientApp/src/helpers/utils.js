import queryString from "query-string";

/**
 * Create a query string out of an object
 * @param obj the key-value pairs that you'd like to be turned into a querystring
 * @returns {string}
 */
export function objectToQueryString(obj) {
	let str = [];
	for (let p in obj)
		if (obj.hasOwnProperty(p)) {
			str.push(encodeURIComponent(p) + "=" + encodeURIComponent(obj[p]));
		}
	if (str.length) return "?" + str.join("&");
	return "";
}

export function queryStringToObject(qryString) {
	return queryString.parse(qryString);
}

/**
 * Resolve a promise on the next tick - used in testing
 * @see https://nodejs.org/en/docs/guides/event-loop-timers-and-nexttick/#process-nexttick
 * @returns {Promise<any>}
 */
export const nextTick = async () => {
	return new Promise((resolve) => {
		setTimeout(resolve, 0);
	});
};

/**
 * Format a string by replacing {0} with arguments passed in
 * @see https://stackoverflow.com/a/4673436
 * @returns {String}
 */
export function replaceFormat(stringToReplace, args) {
	if (typeof(stringToReplace) === undefined || stringToReplace === null)
		return stringToReplace;

	return stringToReplace.replace(/{(\d+)}/g, function(match, number) {
		return typeof args[number] !== undefined
			? args[number]
			: match
		;
	});
}

/**
 * @param String The string in which you want to check for the presence of "http"
 * Detects link that starts with "http"
 * @returns {Boolean}
 */
export const isHttpLink = link => link.indexOf("http") === 0;
