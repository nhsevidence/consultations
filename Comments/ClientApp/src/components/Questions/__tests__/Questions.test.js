/* global jest */

import React from "react";
import { mount } from "enzyme";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import { MemoryRouter } from "react-router";
import { nextTick } from "../../../helpers/utils";
import toJson from "enzyme-to-json";
import QuestionsData from "./QuestionsData.json";
import { Questions, AddQuestionButton } from "../Questions";
import { OpenQuestion } from "../../QuestionTypes/OpenQuestion/OpenQuestion";

const textQuestionId = QuestionsData.questionTypes[0].questionTypeId;
const yesNoQuestionId = QuestionsData.questionTypes[1].questionTypeId;

const mock = new MockAdapter(axios);

jest.mock("../../../context/UserContext", () => {
	return {
		UserContext: {
			Consumer: (props) => {
				return props.children({
					isAuthorised: true,
				});
			},
		},
	};
});

describe("[ClientApp] ", () => {
	describe("Question administration Component", () => {

		const fakeProps = {
			match: {
				params: {
					consultationId: 1,
				},
			},
			location: {
				pathname: "/admin/questions/1",
			},
		};

		beforeEach(() => {
			mock
				.onGet()
				.reply(200, QuestionsData);
		});

		afterEach(() => {
			mock.reset();
		});

		it("should match the supplied snapshot", async () => {
			const wrapper = mount(
				<MemoryRouter>
					<Questions {...fakeProps}/>
				</MemoryRouter>,
			);

			await nextTick();
			wrapper.update();

			expect(
				toJson(wrapper, {
					noKey: true,
					mode: "deep",
				}),
			).toMatchSnapshot();
		});

		it("should not render an add question button at the default URL (e.g. 'questions/1)", async () => {
			const wrapper = mount(
				<MemoryRouter>
					<Questions {...fakeProps}/>
				</MemoryRouter>,
			);

			await nextTick();
			wrapper.update();

			expect(wrapper.find(<OpenQuestion/>).length).toEqual(0);
		});

		it("should render a single consultation question", async () => {
			const localProps = Object.assign({}, fakeProps);
			localProps.match.params.documentId = "consultation";

			const wrapper = mount(
				<MemoryRouter>
					<Questions {...localProps}/>
				</MemoryRouter>,
			);

			await nextTick();
			wrapper.update();
			const length = wrapper.find(OpenQuestion).length;

			expect(length).toEqual(1);
		});

		it("should render a marker with the correct quantity of questions", async () => {
			const localProps = Object.assign({}, fakeProps);
			localProps.match.params.documentId = "consultation";

			const wrapper = mount(
				<MemoryRouter>
					<Questions {...localProps}/>
				</MemoryRouter>,
			);

			await nextTick();
			wrapper.update();

			const text = wrapper.find("[aria-current='page']").find("span.text-right").text();
			expect(text).toEqual("(1)");
		});

		it("should increment the question count marker when the add question button is clicked ", async () => {
			const localProps = Object.assign({}, fakeProps);
			localProps.match.params.documentId = "consultation";

			const wrapper = mount(
				<MemoryRouter>
					<Questions {...localProps}/>
				</MemoryRouter>,
			);

			await nextTick();
			wrapper.update();

			expect(wrapper.find("[aria-current='page']").find("span.text-right").text()).toEqual("(1)");

			wrapper.find(".btn--cta").first().simulate("click");

			await nextTick();
			wrapper.update();

			expect(wrapper.find("[aria-current='page']").find("span.text-right").text()).toEqual("(2)");
		});

		it("get consultation level questions", () => {
			const thing = new Questions();
			const result = thing.getQuestionsToDisplay(
				"consultation",
				QuestionsData,
			);
			expect(result).toEqual(QuestionsData.consultationQuestions);
		});

		it("get document level questions", () => {
			const thing = new Questions();
			const result = thing.getQuestionsToDisplay(
				"1",
				QuestionsData,
			);
			expect(result).toEqual(QuestionsData.documents[0].documentQuestions);
		});

		it("should add a blank open question to state when the add open question button is clicked", async () => {
			const localProps = Object.assign({}, fakeProps);
			localProps.match.params.documentId = "consultation";
			const wrapper = mount(<MemoryRouter><Questions {...localProps}/></MemoryRouter>);
			await nextTick();
			wrapper.update();
			const questionInstance = wrapper.find(Questions).instance();
			expect(questionInstance.state.questionsData.consultationQuestions.length).toEqual(1);
			const textQuestionButton = wrapper.find("button[children='Add an open question']");
			textQuestionButton.simulate("click");
			await nextTick();
			wrapper.update();
			expect(questionInstance.state.questionsData.consultationQuestions.length).toEqual(2);
			expect(questionInstance.state.questionsData.consultationQuestions[1].questionTypeId).toEqual(textQuestionId);
		});

		it("should add a yes / no question when the yes / no button is clicked", async () => {
			const localProps = Object.assign({}, fakeProps);
			localProps.match.params.documentId = "consultation";
			const wrapper = mount(<MemoryRouter><Questions {...localProps}/></MemoryRouter>);
			await nextTick();
			wrapper.update();
			const questionInstance = wrapper.find(Questions).instance();
			const textQuestionButton = wrapper.find("button[children='Add a yes/no question']");
			textQuestionButton.simulate("click");
			await nextTick();
			wrapper.update();
			expect(questionInstance.state.questionsData.consultationQuestions[1].questionTypeId).toEqual(yesNoQuestionId);
		});

		it("should render an add question button for each of the question types in the data", async () => {
			const localProps = Object.assign({}, fakeProps);
			localProps.match.params.documentId = "consultation";
			const wrapper = mount(<MemoryRouter><Questions {...localProps}/></MemoryRouter>);
			await nextTick();
			wrapper.update();
			expect(wrapper.find(AddQuestionButton).length).toEqual(QuestionsData.questionTypes.length);
		});

	});
});
