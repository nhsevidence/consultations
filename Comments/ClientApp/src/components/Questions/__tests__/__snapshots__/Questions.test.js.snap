// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ClientApp]  Question administration Component should match the supplied snapshot 1`] = `null`;
